src/Http/Routing/src/Constraints/RegexRouteConstraint.cs

```diff
@@ -27,20 +28,24 @@ public RegexRouteConstraint(Regex regex)
         ArgumentNullException.ThrowIfNull(regex);
 
         _constraint = regex;
-        _regexPattern = regex.ToString();
     }
 
     /// <summary>
     /// Constructor for a <see cref="RegexRouteConstraint"/> given a <paramref name="regexPattern"/>.
     /// </summary>
     /// <param name="regexPattern">A string containing the regex pattern.</param>
     public RegexRouteConstraint(
-        [StringSyntax(StringSyntaxAttribute.Regex, RegexOptions.CultureInvariant | RegexOptions.IgnoreCase)]
+        [StringSyntax(StringSyntaxAttribute.Regex, RegexOptions.CultureInvariant | RegexOptions.Compiled | RegexOptions.IgnoreCase)]
         string regexPattern)
     {
         ArgumentNullException.ThrowIfNull(regexPattern);
 
-        _regexPattern = regexPattern;
+        // Create regex instance lazily to avoid compiling regexes at app startup. Delay creation until Constraint is first evaluated.
+        // This is not thread safe. No side effect but multiple instances of a regex instance could be created from a burst of requests.
```

Comment:
```suggestion
        // This is not thread-safe. No side effect, but multiple instances of a regex instance could be created from a burst of requests.
        // Regex instance is created in a func to avoid bringing in the regex engine when RegexRouteConstraint is created from a source-generated regex.
```